common/quotas/global/doc.go

```diff
@@ -0,0 +1,161 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+/*
+Package global contains a global-load-balance-aware ratelimiter (when complete).
+
+# High level overview
+
+At a very high level, this will:
+  - collect usage metrics in-memory in "limiting" hosts (which use an in-memory ratelimiter)
+  - asynchronously submit these usage metrics to "aggregating" hosts, which will return per-key RPS data
+  - "limiting" hosts use this returned value to update their in-memory ratelimiters
+
+"Aggregating" hosts will internally:
+  - receive data for key X from all "limiting" hosts
+  - compute request-weight per host+key in the cluster
+  - return the "limiting"-host's weighted portion of each ratelimit to allow
+
+The exact logic that the "aggregating" hosts use is intentionally hidden from the
+"limiting" hosts, so it can be changed without changing how they enforce limits.
+
+The weight-calculation algorithm will (eventually) be controlled by a dynamicconfig value,
+to allow shadowing and experimenting with different algorithms at runtime, or disabling
+the global logic altogether.
+
+See sub-packages for implementation details.
+
+# Data flow in the system
+
+The overall planned system's data-flow looks roughly like this:
+
+	┌───────┐
+	│users  │
+	└┬─┬─┬─┬┘
+	 a b c c
+	 │ │ │ │
+	┌▽─▽─▽─▽───┐
+	│frontend  │        (rejects based on previous or fallback limits, collects call counts)
+	│(limiting)│
+	└───┬──────┘
+	    │
+	 [a1,b1,c2]         (sends aggregated usage data to agg hosts, split by ring on keys)
+	    │
+	┌───▽─────────┐
+	│ring         │     (ringpop `Lookup(key)` to find agg host per key)
+	└┬─────┬─────┬┘
+	 │     │     │
+	 a1    b1    c2     (each subset of keys is sent to its owning agg host)
+	 │     │     │
+	 │     │     │  ┌─────── other "c" usage from other frontends, etc
+	┌▽───┐┌▽───┐┌▽──▽┐
+	│agg1││agg2││agg3│  (aggs compare frontend's usage per key with others, decide on RPS to allow)
+	└┬───┘└┬───┘└┬───┘
+	 │     │     │
+	a/2   b/4   c3/4    (aggs return weighted RPS to allow)
```

Comment:
ensuring I'm interpreting this correctly, this is a push or pull model? Do the frontend hosts poll the history host aggregators or does the aggregator reach out with a notion of subscription?

Reply:
frontends poll / repeatedly call history aggregators, those are the only requests.

I was trying to figure out how to make that clear from the diagram, but kinda gave up.  everything I can come up with needs more width to draw return-path lines, and that just gets worse and worse to render and display :\

Reply:
did a plain sequence diagram, and I think it might be better (updating in a moment)